src/generateDiff.js

```diff
@@ -11,30 +18,64 @@ const diff = require('diff');
  */
 function generateDiff(before, after) {
   const changes = diff.diffLines(before, after);
-  const diffLines = [];
-  const preceedingContext = [];
-  for (const { added, removed, value } of changes) {
-    const lines = value.split('\n');
-    // remove trailing newline
-    lines.pop();
-    if (added || removed) {
-      if (preceedingContext.length) {
-        diffLines.push(...preceedingContext);
-        preceedingContext.splice(0, preceedingContext.length);
-      }
-      diffLines.push(...lines.map((line) => `${added ? '+' : '-'}${line}`));
-    } else {
-      // If a changed line has been included already, add up to 2 lines of context
-      if (diffLines.length) {
-        diffLines.push(...lines.slice(0, 2).map((line) => ` ${line}`));
-        lines.splice(0, 2);
+  // `diffLines` will always return at least one change
+  const lastChange = changes[changes.length - 1];
+  const penultimateChange = changes[changes.length - 2] || {};
+
+  // Add notice about newline at end of file
+  if (!lastChange.value.endsWith('\n')) {
+    lastChange.noNewline = true;
+  }
+  // If the last change is an addition and the penultimate change is a
+  // removal, then the last line of the file is also in the penultimate change.
+  // That's why we're checking to see if the newline notice is needed here as
+  // well.
+  if (
+    lastChange.added &&
+    penultimateChange.removed &&
+    !penultimateChange.value.endsWith('\n')
+  ) {
+    penultimateChange.noNewline = true;
+  }
+
+  const diffLines = changes.flatMap(
+    ({ added, noNewline, removed, value }, index) => {
+      const lines = getTrimmedLines(value);
+      const changedLines = [];
+      if (added || removed) {
+        // Add up to 2 lines of context before each change
+        const previousContext = changes[index - 1];
+        if (
+          previousContext &&
+          !previousContext.added &&
+          !previousContext.removed
+        ) {
+          const hasPreviousChange = index > 1;
```

Comment:
This could use some more explanation, either be a comment here or expanding the comment on L21 above.